### PR TITLE
feat(scim): add SCIM filter expression parser with unit tests

### PR DESCRIPTION
**Pull request by nmgarza5**
## Description

Add a SCIM filter expression parser (`filtering.py`) that handles the subset of RFC 7644 filter syntax that identity providers (Okta, Azure AD) actually send in practice. Supports `eq`, `co`, and `sw` operators on any attribute.

Includes comprehensive unit tests covering quoted values, case-insensitive operators, edge cases, and malformed input.

Linear: https://linear.app/onyx-app/issue/ENG-3643

## Additional Options

- [x] [Required] I have considered whether this PR needs to be cherry-picked to the latest beta branch.
- [ ] [Optional] Override Linear Check

---
## Summary by cubic
Add a SCIM filter parser for simple IdP filters with unit tests to support reliable SCIM lookups. Addresses ENG-3643.

- **New Features**
  - Supported operators: eq, co, sw; operators are case-insensitive.
  - Accepts single- or double-quoted values; trims surrounding whitespace.
  - Empty/missing filter returns None; malformed, unquoted, unsupported, or compound filters raise ValueError.
  - Adds a module docstring clarifying why IdPs use SCIM filters.

<sup>Written for commit 785879cd394643a68c246a7d6e09ffd290c981be. Summary will update on new commits.</sup>

### CODE REVIEW

File: backend/ee/onyx/server/scim/filtering.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+class ScimFilterOp(str, Enum):
+    """Supported SCIM filter operators."""
+
+    EQ = "eq"
+    CO = "co"
+    SW = "sw"
+
+
+@dataclass(frozen=True, slots=True)
```
[the remark this answers is not in the document]
Comment (nmgarza5) [reply backend/ee/onyx/server/scim/filtering.py:39]: Intentionally using `@dataclass(frozen=True, slots=True)` here. `ScimFilter` is a simple, immutable value object used only internally by the parser — it doesn't need Pydantic's validation, serialization, or schema generation. The frozen dataclass is lighter and faster for this use case.

File: backend/tests/unit/onyx/server/scim/test_filtering.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    def test_empty_input_returns_none(self, filter_string: str | None) -> None:
+        assert parse_scim_filter(filter_string) is None
+
+    @pytest.mark.parametrize(
+        "filter_string",
+        [
+            "userName",  # missing operator and value
+            "userName eq",  # missing value
+            'userName gt "5"',  # unsupported operator
+            'userName ne "test"',  # unsupported operator
```
[the remark this answers is not in the document]
Comment (nmgarza5) [reply backend/tests/unit/onyx/server/scim/test_filtering.py:86]: Intentional — we only support operators actually used by Okta and Azure AD for SCIM provisioning (`eq`, `co`, `sw`). `ne` is part of the SCIM spec but isn't sent by any IdP we're targeting. We'll add more operators if needed when we hit real-world IdP requirements.

File: backend/tests/unit/onyx/server/scim/test_filtering.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    @pytest.mark.parametrize(
+        "filter_string",
+        [
+            None,
+            "",
+            "   ",
+        ],
+    )
+    def test_empty_input_returns_none(self, filter_string: str | None) -> None:
+        assert parse_scim_filter(filter_string) is None
```
[the remark this answers is not in the document]
Comment (nmgarza5) [reply backend/tests/unit/onyx/server/scim/test_filtering.py:78]: Fixed — malformed filters now raise `ValueError`. Empty/missing input still returns `None` (no filter = return all).

File: backend/ee/onyx/server/scim/filtering.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+    Returns:
+        A ``ScimFilter`` if the expression is valid and uses a supported
+        operator, otherwise ``None``.
+    """
+    if not filter_string or not filter_string.strip():
+        return None
+
+    match = _FILTER_RE.match(filter_string.strip())
+    if not match:
```
[the remark this answers is not in the document]
Comment (nmgarza5) [reply backend/ee/onyx/server/scim/filtering.py:78]: Fixed in the same commit — all malformed filter paths now raise `ValueError`.

File: backend/ee/onyx/server/scim/filtering.py [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+
+    attribute: str
+    operator: ScimFilterOp
+    value: str
+
+
+# Matches: attribute operator "value" (with or without quotes around value)
+# Groups: (attribute) (operator) ("quoted value" | unquoted_value)
+_FILTER_RE = re.compile(
+    r"^(\S+)\s+(eq|co|sw)\s+"  # attribute + operator
```
[the remark this answers is not in the document]
Comment (nmgarza5) [reply backend/ee/onyx/server/scim/filtering.py:51]: Same as above — we intentionally support only `eq`, `co`, `sw`. These cover all filter expressions sent by Okta and Azure AD during SCIM provisioning. `ne`, `ew`, and `pr` are in the SCIM spec but aren't used by these IdPs. Unsupported operators now raise `ValueError` rather than silently returning `None`.

File: backend/ee/onyx/server/scim/filtering.py [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+"""
+
+from __future__ import annotations
+
+import re
+from dataclasses import dataclass
+from enum import Enum
+
+
+class ScimFilterOp(str, Enum):
```
Comment (justin-tahara) [inline backend/ee/onyx/server/scim/filtering.py]: Let's spell out like before

File: backend/ee/onyx/server/scim/filtering.py [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+    CO = "co"
+    SW = "sw"
+
+
+@dataclass(frozen=True, slots=True)
+class ScimFilter:
+    """Parsed SCIM filter expression."""
+
+    attribute: str
+    operator: ScimFilterOp
```
Comment (justin-tahara) [inline backend/ee/onyx/server/scim/filtering.py]: Same as above

File: backend/ee/onyx/server/scim/filtering.py [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+        return None
+
+    match = _FILTER_RE.match(filter_string.strip())
+    if not match:
+        raise ValueError(f"Unsupported or malformed SCIM filter: {filter_string}")
+
+    attribute = match.group(1)
+    op_str = match.group(2).lower()
+    # Value is in group 3 (double-quoted) or group 4 (single-quoted)
+    value = match.group(3) if match.group(3) is not None else match.group(4)
```
Comment (justin-tahara) [inline backend/ee/onyx/server/scim/filtering.py:89]: Almost feels like it should be it's own little function. Maybe for testability

**Review by justin-tahara — COMMENTED** [review]
Some comments

File: backend/ee/onyx/server/scim/filtering.py (hunk 6 above)
Comment (nmgarza5) [reply backend/ee/onyx/server/scim/filtering.py]: da fuq....my bad...
